File: 2024.1/lab1/java/concurrent/Sum.java
import java.io.FileInputStream;
import java.io.IOException;
import java.nio.file.Files;
import java.nio.file.Path;
import java.nio.file.Paths;
import java.util.Arrays;

public class Sum {

    public static int sum(FileInputStream fis) throws IOException {
        
	int byteRead;
        int sum = 0;
        
        while ((byteRead = fis.read()) != -1) {
        	sum += byteRead;
        }

        return sum;
    }

    public static long sum(String path) throws IOException {

        Path filePath = Paths.get(path);
        if (Files.isRegularFile(filePath)) {
       	    FileInputStream fis = new FileInputStream(filePath.toString());
            return sum(fis);
        } else {
            throw new RuntimeException("Non-regular file: " + path);
        }
    }

    public static void main(String[] args) throws Exception {

        var numThreads = 4;
        Long[] res = new Long[numThreads];
        int count = 25;
        Thread[] threads = new Thread[numThreads];
        

        for(int i = 0; i < numThreads; i++) {
            int[] size = new int[]{i * count, i * count + count};
            Task t = new Sum(). new Task(res, i, size);
            threads[i] = t;
            threads[i].start();
        }

        for(Thread t : threads) {
            t.join();
        }

        System.out.println(Arrays.toString(res));
    
        Long ccount = 0L;
        for(Long value : res){
            ccount += value;
        }

        System.out.println(ccount);

    }


    class Task extends Thread {
        private final Long[] res;
        private final int thread_num;
        private final int[] size;

        public Task(Long[] res, int thread_num, int[] size) {
            this.res = res;
            this.thread_num = thread_num;
            this.size = size;
        }
        
        @Override
        public void run(){
            try {
                res[this.thread_num] = sum_thread();
            } catch (IOException e) {
            }
        }

        public Long sum_thread() throws IOException{
            Long _sum = 0L;

            for(int i = size[0]; i < size[1]; i++){
                _sum += sum("../../dataset/" + "file." + (i + 1));
            }


            return _sum;
        }
    }


}

File: 2024.1/lab1/python/concurrent/sum.py
import os, os.path
import threading
import sys
mutex = threading.Semaphore(1)

def sum_thread(thread_idx, size, res):
    for i in range(size[0], size[1]):
        mutex.acquire()
        res[thread_idx] += do_sum("../../dataset/" + "file." + str(i  + 1))
        mutex.release()

def do_sum(path):
    _sum = 0
    with open(path, 'rb',buffering=0) as f:
        byte = f.read(1)
        while byte:
            _sum += int.from_bytes(byte, byteorder='big', signed=False)
            byte = f.read(1)
        return _sum

def count_files():
    return len(os.listdir('../../dataset/'))

if __name__ == "__main__":
    num_threads = 4
    res = [0] * num_threads
    threads = []
    count = count_files() // num_threads
    for i in range(num_threads):
        t = threading.Thread(target =sum_thread, args=(i, (i * count,
            i * count + count), res))
        threads.append(t)
        t.start()

    for t in threads:
        t.join()

    print(res)

    print(sum(res))